src/main/java/com/es/phoneshop/model/cart/DefaultCartService.java

```diff
@@ -31,25 +33,63 @@ public synchronized Cart getCart(HttpSession session) {
     }
 
     @Override
-    public synchronized void add(Cart cart, Long productId, int quantity)  throws OutOfStockException {
-        if (quantity <= 0) {
-            throw new OutOfStockException(null, quantity, 0);
+    public synchronized void add(Cart cart, Long productId, int quantity) throws OutOfStockException {
+        Product product = productDao.getProduct(productId);
+        Optional<CartItem> cartItemOptional = getCartItemOptional(cart, product);
+        int productsAmount = cartItemOptional.map(CartItem::getQuantity).orElse(0);
+
+        checkQuantity(quantity, productsAmount + quantity, product);
+
+        if (cartItemOptional.isPresent()) {
+            cartItemOptional.get().setQuantity(productsAmount + quantity);
+        } else {
+            cart.getItems().add(new CartItem(product, quantity));
         }
+        updateMiniCart(cart);
+    }
 
+    @Override
+    public synchronized void update(Cart cart, Long productId, int quantity) throws OutOfStockException {
         Product product = productDao.getProduct(productId);
-        Optional<CartItem> cartItemOptional = cart.getItems().stream()
+        Optional<CartItem> cartItemOptional = getCartItemOptional(cart, product);
+
+        checkQuantity(quantity, quantity, product);
+
+        cartItemOptional.ifPresent(cartItem -> cartItem.setQuantity(quantity));
+        updateMiniCart(cart);
+    }
+
+    private Optional<CartItem> getCartItemOptional(Cart cart, Product product) {
+        return cart.getItems().stream()
                 .filter(c -> product.getId().equals(c.getProduct().getId()))
                 .findAny();
-        int productsAmount = cartItemOptional.map(CartItem::getQuantity).orElse(0);
+    }
 
-        if (product.getStock() < productsAmount + quantity) {
-            throw new OutOfStockException(product, productsAmount + quantity, product.getStock());
+    private void checkQuantity(int quantity, int newQuantity, Product product) throws OutOfStockException {
+        if (quantity <= 0) {
+            throw new OutOfStockException(null, quantity, 0);
+        }
+        if (product.getStock() < newQuantity) {
+            throw new OutOfStockException(product, newQuantity, product.getStock());
         }
+    }
 
+    @Override
+    public synchronized void delete(Cart cart, Long productId) {
+        Product product = productDao.getProduct(productId);
+        Optional<CartItem> cartItemOptional = getCartItemOptional(cart, product);
         if (cartItemOptional.isPresent()) {
-            cartItemOptional.get().setQuantity(productsAmount + quantity);
-        } else {
-            cart.getItems().add(new CartItem(product, quantity));
+            cart.getItems().removeIf(cartItem -> productId.equals(cartItem.getProduct().getId()));
         }
+        updateMiniCart(cart);
+    }
+
+    private void updateMiniCart(Cart cart) {
```

Comment:
Rename it to smth like recalculateCart, it doesn't really have any relation to minicart (which is only a visual representation of a user's cart)